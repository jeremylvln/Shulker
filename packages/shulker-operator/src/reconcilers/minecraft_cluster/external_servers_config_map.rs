use std::collections::BTreeMap;

use k8s_openapi::api::core::v1::ConfigMap;
use kube::core::ObjectMeta;
use kube::Api;
use kube::Client;
use kube::ResourceExt;

use shulker_crds::v1alpha1::minecraft_cluster::MinecraftCluster;
use shulker_kube_utils::reconcilers::builder::ResourceBuilder;

use super::MinecraftClusterReconciler;

pub struct ExternalServersConfigMapBuilder {
    client: Client,
}

#[async_trait::async_trait]
impl<'a> ResourceBuilder<'a> for ExternalServersConfigMapBuilder {
    type OwnerType = MinecraftCluster;
    type ResourceType = ConfigMap;
    type Context = ();

    fn name(cluster: &Self::OwnerType) -> String {
        format!("{}-external-servers", cluster.name_any())
    }

    fn api(&self, cluster: &Self::OwnerType) -> kube::Api<Self::ResourceType> {
        Api::namespaced(self.client.clone(), cluster.namespace().as_ref().unwrap())
    }

    fn is_needed(&self, cluster: &Self::OwnerType) -> bool {
        cluster
            .spec
            .external_servers
            .as_ref()
            .map_or(false, |list| !list.is_empty())
    }

    async fn build(
        &self,
        cluster: &Self::OwnerType,
        name: &str,
        _existing_config_map: Option<&Self::ResourceType>,
        _context: Option<Self::Context>,
    ) -> Result<Self::ResourceType, anyhow::Error> {
        let config_map = ConfigMap {
            metadata: ObjectMeta {
                name: Some(name.to_string()),
                namespace: Some(cluster.namespace().unwrap().clone()),
                labels: Some(MinecraftClusterReconciler::get_labels(
                    cluster,
                    "external-servers".to_string(),
                    "proxy".to_string(),
                )),
                ..ObjectMeta::default()
            },
            data: Some(BTreeMap::from([(
                "external-servers.yaml".to_string(),
                ExternalServersConfigMapBuilder::get_content_from_server_list(cluster),
            )])),
            ..ConfigMap::default()
        };

        Ok(config_map)
    }
}

impl ExternalServersConfigMapBuilder {
    pub fn new(client: Client) -> Self {
        ExternalServersConfigMapBuilder { client }
    }

    fn get_content_from_server_list(cluster: &MinecraftCluster) -> String {
        serde_yaml::to_string(cluster.spec.external_servers.as_ref().unwrap()).unwrap()
    }
}

#[cfg(test)]
mod tests {
    use shulker_kube_utils::reconcilers::builder::ResourceBuilder;

    use crate::reconcilers::minecraft_cluster::fixtures::{create_client_mock, TEST_CLUSTER};

    #[test]
    fn name_contains_cluster_name() {
        // W
        let name = super::ExternalServersConfigMapBuilder::name(&TEST_CLUSTER);

        // T
        assert_eq!(name, "my-cluster-external-servers");
    }

    #[tokio::test]
    async fn build_snapshot() {
        // G
        let client = create_client_mock();
        let builder = super::ExternalServersConfigMapBuilder::new(client);
        let name = super::ExternalServersConfigMapBuilder::name(&TEST_CLUSTER);

        // W
        let config_map = builder
            .build(&TEST_CLUSTER, &name, None, None)
            .await
            .unwrap();

        // T
        insta::assert_yaml_snapshot!(config_map);
    }
}
