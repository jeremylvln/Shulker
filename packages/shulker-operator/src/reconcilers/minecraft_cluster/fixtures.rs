use http::{Request, Response};
use kube::client::Body;
use kube::{core::ObjectMeta, Client};
use lazy_static::lazy_static;
use shulker_crds::v1alpha1::minecraft_cluster::{
    MinecraftCluster, MinecraftClusterExternalServerSpec, MinecraftClusterSpec,
};

lazy_static! {
    pub static ref TEST_CLUSTER: MinecraftCluster = MinecraftCluster {
        metadata: ObjectMeta {
            namespace: Some("default".to_string()),
            name: Some("my-cluster".to_string()),
            ..ObjectMeta::default()
        },
        spec: MinecraftClusterSpec {
            network_admins: None,
            redis: None,
            external_servers: Some(vec![MinecraftClusterExternalServerSpec {
                name: "my-external-server".to_string(),
                address: "127.0.0.1:25565".to_string(),
                tags: vec!["game".to_string()]
            }])
        },
        status: None,
    };
}

pub fn create_client_mock() -> Client {
    let (mock_service, _) = tower_test::mock::pair::<Request<Body>, Response<Body>>();
    Client::new(mock_service, "default")
}
