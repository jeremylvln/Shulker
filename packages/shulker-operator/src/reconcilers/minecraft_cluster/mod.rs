use std::{collections::BTreeMap, sync::Arc, time::Duration};

use external_servers_config_map::ExternalServersConfigMapBuilder;
use futures::StreamExt;
use k8s_openapi::api::{
    apps::v1::StatefulSet,
    core::v1::{ConfigMap, Secret, Service, ServiceAccount},
    rbac::v1::{Role, RoleBinding},
};
use kube::{
    api::ListParams,
    runtime::{
        controller::Action,
        finalizer::{finalizer, Event as Finalizer},
        watcher::Config,
        Controller,
    },
    Api, Client, ResourceExt,
};
use shulker_kube_utils::reconcilers::builder::reconcile_builder;
use tracing::*;

use shulker_crds::v1alpha1::minecraft_cluster::MinecraftCluster;

use crate::reconcilers::ReconcilerError;

use self::{
    forwarding_secret::ForwardingSecretBuilder, headless_service::HeadlessServiceBuilder,
    minecraft_server_role::MinecraftServerRoleBuilder,
    minecraft_server_role_binding::MinecraftServerRoleBindingBuilder,
    minecraft_server_service_account::MinecraftServerServiceAccountBuilder,
    proxy_role::ProxyRoleBuilder, proxy_role_binding::ProxyRoleBindingBuilder,
    proxy_service_account::ProxyServiceAccountBuilder, redis_service::RedisServiceBuilder,
    redis_stateful_set::RedisStatefulSetBuilder,
};

use super::Result;

pub mod external_servers_config_map;
mod forwarding_secret;
mod headless_service;
mod minecraft_server_role;
mod minecraft_server_role_binding;
mod minecraft_server_service_account;
mod proxy_role;
mod proxy_role_binding;
mod proxy_service_account;
pub mod redis_service;
mod redis_stateful_set;

#[cfg(test)]
pub mod fixtures;

static FINALIZER: &str = "minecraftclusters.shulkermc.io";

struct MinecraftClusterReconciler {
    client: kube::Client,

    // Builders
    forwarding_secret_builder: ForwardingSecretBuilder,
    headless_service_builder: HeadlessServiceBuilder,
    proxy_service_account_builder: ProxyServiceAccountBuilder,
    proxy_role_builder: ProxyRoleBuilder,
    proxy_role_binding_builder: ProxyRoleBindingBuilder,
    minecraft_server_service_account_builder: MinecraftServerServiceAccountBuilder,
    minecraft_server_role_builder: MinecraftServerRoleBuilder,
    minecraft_server_role_binding_builder: MinecraftServerRoleBindingBuilder,
    redis_service_builder: RedisServiceBuilder,
    redis_stateful_set_builder: RedisStatefulSetBuilder,
    external_servers_config_map_builder: ExternalServersConfigMapBuilder,
}

impl MinecraftClusterReconciler {
    async fn reconcile(
        &self,
        _api: Api<MinecraftCluster>,
        cluster: Arc<MinecraftCluster>,
    ) -> Result<Action> {
        reconcile_builder(&self.forwarding_secret_builder, cluster.as_ref(), None)
            .await
            .map_err(ReconcilerError::BuilderError)?;
        reconcile_builder(&self.headless_service_builder, cluster.as_ref(), None)
            .await
            .map_err(ReconcilerError::BuilderError)?;
        reconcile_builder(&self.proxy_service_account_builder, cluster.as_ref(), None)
            .await
            .map_err(ReconcilerError::BuilderError)?;
        reconcile_builder(&self.proxy_role_builder, cluster.as_ref(), None)
            .await
            .map_err(ReconcilerError::BuilderError)?;
        reconcile_builder(&self.proxy_role_binding_builder, cluster.as_ref(), None)
            .await
            .map_err(ReconcilerError::BuilderError)?;
        reconcile_builder(
            &self.minecraft_server_service_account_builder,
            cluster.as_ref(),
            None,
        )
        .await
        .map_err(ReconcilerError::BuilderError)?;
        reconcile_builder(&self.minecraft_server_role_builder, cluster.as_ref(), None)
            .await
            .map_err(ReconcilerError::BuilderError)?;
        reconcile_builder(
            &self.minecraft_server_role_binding_builder,
            cluster.as_ref(),
            None,
        )
        .await
        .map_err(ReconcilerError::BuilderError)?;
        reconcile_builder(&self.redis_service_builder, cluster.as_ref(), None)
            .await
            .map_err(ReconcilerError::BuilderError)?;
        reconcile_builder(&self.redis_stateful_set_builder, cluster.as_ref(), None)
            .await
            .map_err(ReconcilerError::BuilderError)?;
        reconcile_builder(
            &self.external_servers_config_map_builder,
            cluster.as_ref(),
            None,
        )
        .await
        .map_err(ReconcilerError::BuilderError)?;

        Ok(Action::requeue(Duration::from_secs(5 * 60)))
    }

    async fn cleanup(&self, cluster: Arc<MinecraftCluster>) -> Result<Action> {
        info!(
            name = cluster.name_any(),
            namespace = cluster.namespace(),
            "cleaning up MinecraftCluster",
        );

        Ok(Action::await_change())
    }

    fn get_labels(
        cluster: &MinecraftCluster,
        name: String,
        component: String,
    ) -> BTreeMap<String, String> {
        BTreeMap::from([
            ("app.kubernetes.io/name".to_string(), name.clone()),
            (
                "app.kubernetes.io/instance".to_string(),
                format!("{}-{}", name, cluster.name_any()),
            ),
            ("app.kubernetes.io/component".to_string(), component),
            (
                "app.kubernetes.io/part-of".to_string(),
                format!("cluster-{}", cluster.name_any()),
            ),
            (
                "app.kubernetes.io/managed-by".to_string(),
                "shulker-operator".to_string(),
            ),
            (
                "minecraftcluster.shulkermc.io/name".to_string(),
                cluster.name_any(),
            ),
        ])
    }
}

#[instrument(skip(ctx, cluster))]
async fn reconcile(
    cluster: Arc<MinecraftCluster>,
    ctx: Arc<MinecraftClusterReconciler>,
) -> Result<Action> {
    let ns = cluster.namespace().unwrap();
    let clusters_api: Api<MinecraftCluster> = Api::namespaced(ctx.client.clone(), &ns);

    info!(
        name = cluster.name_any(),
        namespace = ns,
        "reconciling MinecraftCluster",
    );

    finalizer(&clusters_api, FINALIZER, cluster, |event| async {
        match event {
            Finalizer::Apply(cluster) => ctx.reconcile(clusters_api.clone(), cluster.clone()).await,
            Finalizer::Cleanup(cluster) => ctx.cleanup(cluster.clone()).await,
        }
    })
    .await
    .map_err(|e| ReconcilerError::FinalizerError(Box::new(e)))
}

fn error_policy(
    _cluster: Arc<MinecraftCluster>,
    error: &ReconcilerError,
    _ctx: Arc<MinecraftClusterReconciler>,
) -> Action {
    warn!("reconcile failed: {:?}", error);
    // ctx.metrics.reconcile_failure(&cluster, error);
    Action::requeue(Duration::from_secs(5))
}

pub async fn run(client: Client) {
    let clusters_api = Api::<MinecraftCluster>::all(client.clone());
    if let Err(e) = clusters_api.list(&ListParams::default().limit(1)).await {
        error!("CRD is not queryable; {e:?}. Is the CRD installed?");
        std::process::exit(1);
    }

    let context = MinecraftClusterReconciler {
        client: client.clone(),
        forwarding_secret_builder: ForwardingSecretBuilder::new(client.clone()),
        headless_service_builder: HeadlessServiceBuilder::new(client.clone()),
        proxy_service_account_builder: ProxyServiceAccountBuilder::new(client.clone()),
        proxy_role_builder: ProxyRoleBuilder::new(client.clone()),
        proxy_role_binding_builder: ProxyRoleBindingBuilder::new(client.clone()),
        minecraft_server_service_account_builder: MinecraftServerServiceAccountBuilder::new(
            client.clone(),
        ),
        minecraft_server_role_builder: MinecraftServerRoleBuilder::new(client.clone()),
        minecraft_server_role_binding_builder: MinecraftServerRoleBindingBuilder::new(
            client.clone(),
        ),
        redis_service_builder: RedisServiceBuilder::new(client.clone()),
        redis_stateful_set_builder: RedisStatefulSetBuilder::new(client.clone()),

        external_servers_config_map_builder: ExternalServersConfigMapBuilder::new(client.clone()),
    };

    Controller::new(clusters_api, Config::default().any_semantic())
        .owns(
            Api::<ConfigMap>::all(client.clone()),
            Config::default().any_semantic(),
        )
        .owns(
            Api::<Secret>::all(client.clone()),
            Config::default().any_semantic(),
        )
        .owns(
            Api::<ServiceAccount>::all(client.clone()),
            Config::default().any_semantic(),
        )
        .owns(
            Api::<Role>::all(client.clone()),
            Config::default().any_semantic(),
        )
        .owns(
            Api::<RoleBinding>::all(client.clone()),
            Config::default().any_semantic(),
        )
        .owns(
            Api::<Service>::all(client.clone()),
            Config::default().any_semantic(),
        )
        .owns(
            Api::<StatefulSet>::all(client.clone()),
            Config::default().any_semantic(),
        )
        .shutdown_on_signal()
        .run(reconcile, error_policy, context.into())
        .filter_map(|x| async move { std::result::Result::ok(x) })
        .for_each(|_| futures::future::ready(()))
        .await;
}
