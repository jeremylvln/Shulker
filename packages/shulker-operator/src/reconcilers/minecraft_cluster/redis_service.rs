use k8s_openapi::api::core::v1::Service;
use k8s_openapi::api::core::v1::ServicePort;
use k8s_openapi::api::core::v1::ServiceSpec;
use k8s_openapi::apimachinery::pkg::util::intstr::IntOrString;
use kube::core::ObjectMeta;
use kube::Api;
use kube::Client;
use kube::ResourceExt;
use shulker_crds::v1alpha1::minecraft_cluster::MinecraftClusterRedisDeploymentType;

use shulker_crds::v1alpha1::minecraft_cluster::MinecraftCluster;
use shulker_kube_utils::reconcilers::builder::ResourceBuilder;

use super::MinecraftClusterReconciler;

pub struct RedisServiceBuilder {
    client: Client,
}

#[async_trait::async_trait]
impl<'a> ResourceBuilder<'a> for RedisServiceBuilder {
    type OwnerType = MinecraftCluster;
    type ResourceType = Service;
    type Context = ();

    fn name(cluster: &Self::OwnerType) -> String {
        format!("{}-redis-managed", cluster.name_any())
    }

    fn api(&self, cluster: &Self::OwnerType) -> kube::Api<Self::ResourceType> {
        Api::namespaced(self.client.clone(), cluster.namespace().as_ref().unwrap())
    }

    fn is_needed(&self, cluster: &Self::OwnerType) -> bool {
        cluster.spec.redis.as_ref().map_or(true, |redis| {
            redis.type_ == MinecraftClusterRedisDeploymentType::ManagedSingleNode
        })
    }

    async fn build(
        &self,
        cluster: &Self::OwnerType,
        name: &str,
        _existing_service: Option<&Self::ResourceType>,
        _context: Option<Self::Context>,
    ) -> Result<Self::ResourceType, anyhow::Error> {
        let service = Service {
            metadata: ObjectMeta {
                name: Some(name.to_string()),
                namespace: Some(cluster.namespace().unwrap().clone()),
                labels: Some(MinecraftClusterReconciler::get_labels(
                    cluster,
                    "redis".to_string(),
                    "redis".to_string(),
                )),
                ..ObjectMeta::default()
            },
            spec: Some(ServiceSpec {
                selector: Some(MinecraftClusterReconciler::get_labels(
                    cluster,
                    "redis".to_string(),
                    "redis".to_string(),
                )),
                type_: Some("ClusterIP".to_string()),
                ports: Some(vec![ServicePort {
                    name: Some("redis".to_string()),
                    protocol: Some("TCP".to_string()),
                    port: 6379,
                    target_port: Some(IntOrString::String("redis".to_string())),
                    ..ServicePort::default()
                }]),
                ..ServiceSpec::default()
            }),
            ..Service::default()
        };

        Ok(service)
    }
}

impl RedisServiceBuilder {
    pub fn new(client: Client) -> Self {
        RedisServiceBuilder { client }
    }
}

#[cfg(test)]
mod tests {
    use shulker_crds::v1alpha1::minecraft_cluster::{
        MinecraftClusterRedisDeploymentType, MinecraftClusterRedisSpec,
    };
    use shulker_kube_utils::reconcilers::builder::ResourceBuilder;

    use crate::reconcilers::minecraft_cluster::fixtures::{create_client_mock, TEST_CLUSTER};

    #[test]
    fn name_contains_cluster_name() {
        // W
        let name = super::RedisServiceBuilder::name(&TEST_CLUSTER);

        // T
        assert_eq!(name, "my-cluster-redis-managed");
    }

    #[tokio::test]
    async fn is_needed_without_config() {
        // G
        let client = create_client_mock();
        let builder = super::RedisServiceBuilder::new(client);

        // W
        let is_needed = builder.is_needed(&TEST_CLUSTER);

        // T
        assert!(is_needed);
    }

    #[tokio::test]
    async fn is_needed_with_managed_type() {
        // G
        let client = create_client_mock();
        let builder = super::RedisServiceBuilder::new(client);
        let mut cluster = TEST_CLUSTER.clone();
        cluster.spec.redis = Some(MinecraftClusterRedisSpec {
            type_: MinecraftClusterRedisDeploymentType::ManagedSingleNode,
            ..MinecraftClusterRedisSpec::default()
        });

        // W
        let is_needed = builder.is_needed(&cluster);

        // T
        assert!(is_needed);
    }

    #[tokio::test]
    async fn is_needed_with_provided_type() {
        // G
        let client = create_client_mock();
        let builder = super::RedisServiceBuilder::new(client);
        let mut cluster = TEST_CLUSTER.clone();
        cluster.spec.redis = Some(MinecraftClusterRedisSpec {
            type_: MinecraftClusterRedisDeploymentType::Provided,
            ..MinecraftClusterRedisSpec::default()
        });

        // W
        let is_needed = builder.is_needed(&cluster);

        // T
        assert!(!is_needed);
    }

    #[tokio::test]
    async fn build_snapshot() {
        // G
        let client = create_client_mock();
        let builder = super::RedisServiceBuilder::new(client);
        let name = super::RedisServiceBuilder::name(&TEST_CLUSTER);

        // W
        let service = builder
            .build(&TEST_CLUSTER, &name, None, None)
            .await
            .unwrap();

        // T
        insta::assert_yaml_snapshot!(service);
    }
}
