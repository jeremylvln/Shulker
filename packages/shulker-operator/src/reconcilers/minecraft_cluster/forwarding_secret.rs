use std::collections::BTreeMap;

use k8s_openapi::api::core::v1::Secret;
use kube::core::ObjectMeta;
use kube::Api;
use kube::Client;
use kube::ResourceExt;
use rand::distributions::{Alphanumeric, DistString};

use crate::reconcilers::builder::ResourceBuilder;

use shulker_crds::v1alpha1::minecraft_cluster::MinecraftCluster;

use super::MinecraftClusterReconciler;

const SECRET_DATA_KEY: &str = "key";

pub struct ForwardingSecretBuilder {
    client: Client,
}

#[async_trait::async_trait]
impl ResourceBuilder for ForwardingSecretBuilder {
    type OwnerType = MinecraftCluster;
    type ResourceType = Secret;

    fn name(cluster: &Self::OwnerType) -> String {
        format!("{}-forwarding-secret", cluster.name_any())
    }

    fn is_updatable() -> bool {
        false
    }

    fn api(&self, cluster: &Self::OwnerType) -> kube::Api<Self::ResourceType> {
        Api::namespaced(self.client.clone(), cluster.namespace().as_ref().unwrap())
    }

    fn is_needed(&self, _cluster: &Self::OwnerType) -> bool {
        true
    }

    async fn create(
        &self,
        cluster: &Self::OwnerType,
        name: &str,
    ) -> Result<Self::ResourceType, anyhow::Error> {
        let secret = Secret {
            metadata: ObjectMeta {
                name: Some(name.to_string()),
                namespace: Some(cluster.namespace().unwrap().clone()),
                labels: Some(
                    MinecraftClusterReconciler::get_common_labels(cluster)
                        .into_iter()
                        .collect(),
                ),
                ..ObjectMeta::default()
            },
            type_: Some("Opaque".to_string()),
            string_data: Some(BTreeMap::from([(
                SECRET_DATA_KEY.to_string(),
                Self::create_proxy_guard_key(),
            )])),
            ..Secret::default()
        };

        Ok(secret)
    }

    async fn update(
        &self,
        _cluster: &Self::OwnerType,
        _secret: &mut Self::ResourceType,
    ) -> Result<(), anyhow::Error> {
        Ok(())
    }
}

impl ForwardingSecretBuilder {
    pub fn new(client: Client) -> Self {
        ForwardingSecretBuilder { client }
    }

    fn create_proxy_guard_key() -> String {
        let mut rng = rand::thread_rng();
        Alphanumeric.sample_string(&mut rng, 64)
    }
}

#[cfg(test)]
mod tests {
    use crate::reconcilers::{
        builder::ResourceBuilder,
        minecraft_cluster::fixtures::{create_client_mock, TEST_CLUSTER},
    };

    #[test]
    fn name_contains_cluster_name() {
        // W
        let name = super::ForwardingSecretBuilder::name(&TEST_CLUSTER);

        // T
        assert_eq!(name, "my-cluster-forwarding-secret");
    }

    #[tokio::test]
    async fn create_snapshot() {
        // G
        let client = create_client_mock();
        let builder = super::ForwardingSecretBuilder::new(client);

        // W
        let secret = builder
            .create(&TEST_CLUSTER, "my-cluster-forwarding-secret")
            .await
            .unwrap();

        // T
        insta::assert_yaml_snapshot!(secret, {
            ".stringData.key" => "[forwarding secret random redacted]"
        });
    }

    #[tokio::test]
    async fn create_has_secret() {
        // G
        let client = create_client_mock();
        let builder = super::ForwardingSecretBuilder::new(client);

        // W
        let secret = builder
            .create(&TEST_CLUSTER, "my-cluster-forwarding-secret")
            .await
            .unwrap();

        // T
        assert!(secret
            .string_data
            .unwrap()
            .get(super::SECRET_DATA_KEY)
            .is_some());
    }

    #[test]
    fn create_proxy_guard_key_random() {
        // W
        let (secret_1, secret_2, secret_3) = (
            super::ForwardingSecretBuilder::create_proxy_guard_key(),
            super::ForwardingSecretBuilder::create_proxy_guard_key(),
            super::ForwardingSecretBuilder::create_proxy_guard_key(),
        );

        // T
        assert_ne!(secret_1, secret_2);
        assert_ne!(secret_2, secret_3);
        assert_ne!(secret_3, secret_1);
    }
}
