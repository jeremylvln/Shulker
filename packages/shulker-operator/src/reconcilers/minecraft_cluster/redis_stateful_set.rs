use std::collections::BTreeMap;

use k8s_openapi::api::apps::v1::StatefulSet;
use k8s_openapi::api::apps::v1::StatefulSetSpec;
use k8s_openapi::api::core::v1::Capabilities;
use k8s_openapi::api::core::v1::Container;
use k8s_openapi::api::core::v1::ContainerPort;
use k8s_openapi::api::core::v1::EnvVar;
use k8s_openapi::api::core::v1::PersistentVolumeClaim;
use k8s_openapi::api::core::v1::PersistentVolumeClaimSpec;
use k8s_openapi::api::core::v1::PodSpec;
use k8s_openapi::api::core::v1::PodTemplateSpec;
use k8s_openapi::api::core::v1::ResourceRequirements;
use k8s_openapi::api::core::v1::SecurityContext;
use k8s_openapi::api::core::v1::VolumeMount;
use k8s_openapi::apimachinery::pkg::api::resource::Quantity;
use k8s_openapi::apimachinery::pkg::apis::meta::v1::LabelSelector;
use kube::core::ObjectMeta;
use kube::Api;
use kube::Client;
use kube::ResourceExt;
use lazy_static::lazy_static;
use shulker_crds::v1alpha1::minecraft_cluster::MinecraftClusterRedisDeploymentType;

use super::redis_service::RedisServiceBuilder;
use super::MinecraftClusterReconciler;
use crate::reconcilers::builder::ResourceBuilder;
use shulker_crds::v1alpha1::minecraft_cluster::MinecraftCluster;

const REDIS_IMAGE: &str = "redis:7-alpine";
const REDIS_DATA_DIR: &str = "/data";

lazy_static! {
    static ref REDIS_SECURITY_CONTEXT: SecurityContext = SecurityContext {
        allow_privilege_escalation: Some(false),
        read_only_root_filesystem: Some(true),
        run_as_non_root: Some(true),
        run_as_user: Some(1000),
        capabilities: Some(Capabilities {
            drop: Some(vec!["ALL".to_string()]),
            ..Capabilities::default()
        }),
        ..SecurityContext::default()
    };
}

pub struct RedisStatefulSetBuilder {
    client: Client,
}

#[async_trait::async_trait]
impl<'a> ResourceBuilder<'a> for RedisStatefulSetBuilder {
    type OwnerType = MinecraftCluster;
    type ResourceType = StatefulSet;
    type Context = ();

    fn name(cluster: &Self::OwnerType) -> String {
        format!("{}-redis-managed", cluster.name_any())
    }

    fn api(&self, cluster: &Self::OwnerType) -> kube::Api<Self::ResourceType> {
        Api::namespaced(self.client.clone(), cluster.namespace().as_ref().unwrap())
    }

    fn is_needed(&self, cluster: &Self::OwnerType) -> bool {
        cluster.spec.redis.as_ref().map_or(true, |redis| {
            redis.type_ == MinecraftClusterRedisDeploymentType::ManagedSingleNode
        })
    }

    async fn build(
        &self,
        cluster: &Self::OwnerType,
        name: &str,
        _existing_stateful_set: Option<&Self::ResourceType>,
        _context: Option<Self::Context>,
    ) -> Result<Self::ResourceType, anyhow::Error> {
        let stateful_set = StatefulSet {
            metadata: ObjectMeta {
                name: Some(name.to_string()),
                namespace: Some(cluster.namespace().unwrap().clone()),
                labels: Some(MinecraftClusterReconciler::get_labels(
                    cluster,
                    "redis".to_string(),
                    "redis".to_string(),
                )),
                ..ObjectMeta::default()
            },
            spec: Some(StatefulSetSpec {
                selector: LabelSelector {
                    match_labels: Some(MinecraftClusterReconciler::get_labels(
                        cluster,
                        "redis".to_string(),
                        "redis".to_string(),
                    )),
                    ..LabelSelector::default()
                },
                service_name: RedisServiceBuilder::name(cluster),
                replicas: Some(1),
                template: RedisStatefulSetBuilder::get_pod_template_spec(cluster),
                volume_claim_templates: Some(RedisStatefulSetBuilder::get_volume_claim_templates()),
                ..StatefulSetSpec::default()
            }),
            ..StatefulSet::default()
        };

        Ok(stateful_set)
    }
}

impl RedisStatefulSetBuilder {
    pub fn new(client: Client) -> Self {
        RedisStatefulSetBuilder { client }
    }

    fn get_pod_template_spec(cluster: &MinecraftCluster) -> PodTemplateSpec {
        let pod_spec = PodSpec {
            containers: vec![Container {
                image: Some(REDIS_IMAGE.to_string()),
                name: "redis".to_string(),
                command: Some(vec![
                    "redis-server".to_string(),
                    format!("--dir {}", REDIS_DATA_DIR),
                ]),
                env: Some(vec![EnvVar {
                    name: "MASTER".to_string(),
                    value: Some("true".to_string()),
                    ..EnvVar::default()
                }]),
                ports: Some(vec![ContainerPort {
                    name: Some("redis".to_string()),
                    container_port: 6379,
                    ..ContainerPort::default()
                }]),
                security_context: Some(REDIS_SECURITY_CONTEXT.clone()),
                volume_mounts: Some(vec![VolumeMount {
                    name: "data".to_string(),
                    mount_path: REDIS_DATA_DIR.to_string(),
                    ..VolumeMount::default()
                }]),
                ..Container::default()
            }],
            ..PodSpec::default()
        };

        PodTemplateSpec {
            metadata: Some(ObjectMeta {
                labels: Some(MinecraftClusterReconciler::get_labels(
                    cluster,
                    "redis".to_string(),
                    "redis".to_string(),
                )),
                ..ObjectMeta::default()
            }),
            spec: Some(pod_spec),
        }
    }

    fn get_volume_claim_templates() -> Vec<PersistentVolumeClaim> {
        vec![PersistentVolumeClaim {
            metadata: ObjectMeta {
                name: Some("data".to_string()),
                ..ObjectMeta::default()
            },
            spec: Some(PersistentVolumeClaimSpec {
                access_modes: Some(vec!["ReadWriteOnce".to_string()]),
                resources: Some(ResourceRequirements {
                    requests: Some(BTreeMap::from([(
                        "storage".to_string(),
                        Quantity("1Gi".to_string()),
                    )])),
                    ..ResourceRequirements::default()
                }),
                ..PersistentVolumeClaimSpec::default()
            }),
            ..PersistentVolumeClaim::default()
        }]
    }
}

#[cfg(test)]
mod tests {
    use shulker_crds::v1alpha1::minecraft_cluster::{
        MinecraftClusterRedisDeploymentType, MinecraftClusterRedisSpec,
    };

    use crate::reconcilers::{
        builder::ResourceBuilder,
        minecraft_cluster::fixtures::{create_client_mock, TEST_CLUSTER},
    };

    #[test]
    fn name_contains_cluster_name() {
        // W
        let name = super::RedisStatefulSetBuilder::name(&TEST_CLUSTER);

        // T
        assert_eq!(name, "my-cluster-redis-managed");
    }

    #[tokio::test]
    async fn is_needed_without_config() {
        // G
        let client = create_client_mock();
        let builder = super::RedisStatefulSetBuilder::new(client);

        // W
        let is_needed = builder.is_needed(&TEST_CLUSTER);

        // T
        assert!(is_needed);
    }

    #[tokio::test]
    async fn is_needed_with_managed_type() {
        // G
        let client = create_client_mock();
        let builder = super::RedisStatefulSetBuilder::new(client);
        let mut cluster = TEST_CLUSTER.clone();
        cluster.spec.redis = Some(MinecraftClusterRedisSpec {
            type_: MinecraftClusterRedisDeploymentType::ManagedSingleNode,
            ..MinecraftClusterRedisSpec::default()
        });

        // W
        let is_needed = builder.is_needed(&cluster);

        // T
        assert!(is_needed);
    }

    #[tokio::test]
    async fn is_needed_with_provided_type() {
        // G
        let client = create_client_mock();
        let builder = super::RedisStatefulSetBuilder::new(client);
        let mut cluster = TEST_CLUSTER.clone();
        cluster.spec.redis = Some(MinecraftClusterRedisSpec {
            type_: MinecraftClusterRedisDeploymentType::Provided,
            ..MinecraftClusterRedisSpec::default()
        });

        // W
        let is_needed = builder.is_needed(&cluster);

        // T
        assert!(!is_needed);
    }

    #[tokio::test]
    async fn build_snapshot() {
        // G
        let client = create_client_mock();
        let builder = super::RedisStatefulSetBuilder::new(client);
        let name = super::RedisStatefulSetBuilder::name(&TEST_CLUSTER);

        // W
        let role = builder
            .build(&TEST_CLUSTER, &name, None, None)
            .await
            .unwrap();

        // T
        insta::assert_yaml_snapshot!(role);
    }
}
