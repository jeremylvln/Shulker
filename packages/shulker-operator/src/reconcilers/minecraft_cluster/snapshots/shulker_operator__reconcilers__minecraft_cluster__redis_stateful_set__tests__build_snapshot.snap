---
source: packages/shulker-operator/src/reconcilers/minecraft_cluster/redis_stateful_set.rs
expression: role
---
apiVersion: apps/v1
kind: StatefulSet
metadata:
  labels:
    app.kubernetes.io/component: redis
    app.kubernetes.io/instance: redis-my-cluster
    app.kubernetes.io/managed-by: shulker-operator
    app.kubernetes.io/name: redis
    app.kubernetes.io/part-of: cluster-my-cluster
    minecraftcluster.shulkermc.io/name: my-cluster
  name: my-cluster-redis-managed
  namespace: default
spec:
  replicas: 1
  selector:
    matchLabels:
      app.kubernetes.io/component: redis
      app.kubernetes.io/instance: redis-my-cluster
      app.kubernetes.io/managed-by: shulker-operator
      app.kubernetes.io/name: redis
      app.kubernetes.io/part-of: cluster-my-cluster
      minecraftcluster.shulkermc.io/name: my-cluster
  serviceName: my-cluster-redis-managed
  template:
    metadata:
      labels:
        app.kubernetes.io/component: redis
        app.kubernetes.io/instance: redis-my-cluster
        app.kubernetes.io/managed-by: shulker-operator
        app.kubernetes.io/name: redis
        app.kubernetes.io/part-of: cluster-my-cluster
        minecraftcluster.shulkermc.io/name: my-cluster
    spec:
      containers:
        - command:
            - redis-server
            - "--dir /data"
          env:
            - name: MASTER
              value: "true"
          image: "redis:7-alpine"
          name: redis
          ports:
            - containerPort: 6379
              name: redis
          securityContext:
            allowPrivilegeEscalation: false
            capabilities:
              drop:
                - ALL
            readOnlyRootFilesystem: true
            runAsNonRoot: true
            runAsUser: 1000
          volumeMounts:
            - mountPath: /data
              name: data
  volumeClaimTemplates:
    - apiVersion: v1
      kind: PersistentVolumeClaim
      metadata:
        name: data
      spec:
        accessModes:
          - ReadWriteOnce
        resources:
          requests:
            storage: 1Gi

